src/rpc.rs

```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
     }
 }
 
+enum DisconnectorState
+{
+    Connected,
+    Disconnecting,
+    Disconnected
+}
+
+pub struct Disconnector<VatId> where VatId: 'static {
+    connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>,
+    state:  DisconnectorState,
+}
+
+impl <VatId> Disconnector<VatId> {
+    pub fn new(connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>) -> Disconnector<VatId> {
+        let state = match *(connection_state.borrow()) {
+            Some(_) => {
+                DisconnectorState::Connected
+            },
+            None => {
+                DisconnectorState::Disconnected
+            }
+        };
+        Disconnector {
+            connection_state: connection_state,
+            state: state,
+        }
+    }
+    fn disconnect(&self) {
+        if let Some(ref state) = *(self.connection_state.borrow()) {
+            state.disconnect(::capnp::Error::disconnected("client requested disconnect".to_owned()));
+        }
+    }
+}
+
+impl <VatId> Future for Disconnector<VatId>
+where
+    VatId: 'static,
+{
+    type Item = ();
+    type Error = ::capnp::Error;
+
+    fn poll(&mut self) -> ::futures::Poll<Self::Item, Self::Error> {
+        self.state = match self.state {
+            DisconnectorState::Connected => {
+                self.disconnect();
+                DisconnectorState::Disconnecting
+            },
+            DisconnectorState::Disconnecting => {
+                if let Some(_) = *(self.connection_state.borrow()) {
+                    DisconnectorState::Disconnecting
+                } else {
+                    DisconnectorState::Disconnected
+                }
+            },
+            DisconnectorState::Disconnected => {
+                DisconnectorState::Disconnected
+            }
+        };
+        match self.state {
+            DisconnectorState::Connected => unreachable!(),
+            DisconnectorState::Disconnecting => {
+                ::futures::task::current().notify();
```

Comment:
I agree it could be more efficient to only poll again when you know it's going to be disconnected.  However, after calling `disconnect` you need just one turn of the event loop to get the `connection_state` to go away so in practice it's the same.  I thought it was nice that it worked without adding state anywhere else.  If you want to store the task though, who would own it?  `RpcSystem`?  Like, an Rc<Option<Task>> (shared with `Disconnector`) that was along side the `RpcSystem::connection_state` that you'd call `notify` on when you set the `connection_state` to `None`?   I think I prefer to keep it as-is because it's simple and I don't think you'll busy-spin unless there's a bug in `RpcSystem::disconnect`.  Let me know if I'm misunderstanding something though.